### DISCUSSION TITLE
Avoid reallocating a new array when reading it in my `CustomProcedure`

### DISCUSSION THREAD

**Discussion by danielvan1** [microsoft/garnet#665] [Q&A] [answered]
Hi,

I'm trying to avoid reallocating a new array for each time I am reading it in my `CustomProcedure`. I'm working with a loop similar to the one in [Sum.cs](https://github.com/microsoft/garnet/blob/main/main/GarnetServer/Extensions/Sum.cs):

```csharp
Copy code
while ((key = GetNextArg(input, ref offset)).Length > 0) {
    if (garnetApi.GET(ref key, ref ArgSlice array) == GarnetStatus.OK) {
        // Process the array
    }
}
```
From my understanding, the call `garnetApi.GET(ref key, ref ArgSlice array)` allocates a new array each time it's invoked. This interpretation comes from looking at the method in [MainStoreOps.cs](https://github.com/microsoft/garnet/blob/40d9a268cb2aaf7fae0252d27c3205dfa6ec237e/libs/server/Storage/Session/MainStore/MainStoreOps.cs#L82).

I believe that to avoid array reallocation, I should use this method in [MainStoreOps.cs](https://github.com/microsoft/garnet/blob/40d9a268cb2aaf7fae0252d27c3205dfa6ec237e/libs/server/Storage/Session/MainStore/MainStoreOps.cs#L20). 

1. However, I'm unsure about the meaning and usage of the `SpanByte input` parameter in this method. 
2. How do you create the objects `SpanByte input` and `SpanByteAndMemory output` in the method of the `CustomProcedure` 
3. Is this the correct approach to prevent reallocation, or is there a better way to handle this scenario?

Any guidance would be greatly appreciated!

Thanks.

**Comment by yrajas** [comment]
Hi

Thanks for trying this feature.

Using the GET api that takes in the ArgSlice argument is the recommended way to perform GET operations from a custom procedure/transaction as shown in the Sum.cs example. The allocation is required to ensure the input buffer is setup in the right way for the backend processing.

**Reply by danielvan1** [reply]
Sorry for the late reply.

Using the recommended approach I do get an error when working with > 10k arrays each containing 10k doubles:
```
System.OverflowException: Arithmetic operation resulted in an overflow.
   at System.GC.AllocateNewArray(IntPtr typeHandle, Int32 length, GC_ALLOC_FLAGS flags)
   at System.GC.AllocateArray[T](Int32 length, Boolean pinned)
   at Garnet.server.ScratchBufferManager.ExpandScratchBuffer(Int32 newLength)
   at Garnet.server.ScratchBufferManager.FormatScratch(Int32 headerSize, ReadOnlySpan`1 arg)
   at Garnet.server.StorageSession.GET[TContext](ArgSlice key, ArgSlice& value, TContext& context)
   at VectorSum.Execute(IGarnetApi garnetApi, ArgSlice input, MemoryResult`1& output) in VectorSum.cs:line 46
   at Garnet.server.RespServerSession.TryCustomProcedure(Byte id, Byte* ptr, Byte* end, CustomProcedure proc)
   at Garnet.server.RespServerSession.ProcessOtherCommands[TGarnetApi](RespCommand command, TGarnetApi& storageApi)
   at Garnet.server.RespServerSession.ProcessArrayCommands[TGarnetApi](RespCommand cmd, TGarnetApi& storageApi)
   at Garnet.server.RespServerSession.ProcessMessages()
   at Garnet.server.RespServerSession.TryConsumeMessages(Byte* reqBuffer, Int32 bytesReceived)
```

Working with a smaller amount of arrays does work.


The code for the summation looks like this:

```csharp
double sum = new double[10000];
ArgSlice array;
ArgSlice key;
double[] buffer = new double[sum.Length];

while ((key = GetNextArg(input, ref offset)).Length > 0) {
    // SpanByte keySpanByte = key.SpanByte;
    Console.WriteLine($"Key: {key}");

    if (garnetApi.GET(key, out array) == GarnetStatus.OK) {
        Console.WriteLine($"array.Length: {array.Length}");
        serializer.Deserialize(array.Span, ref buffer!);

        for(int i = 0; i < buffer.Length; i++) {
            sum[i] += buffer[i];
        }

    }
}
```

Would you have any idea of how to fix the issue?

**Reply by yrajas** [reply]
Thanks for pointing this out. Looking into it.
Meanwhile, could you please share a repro of the client script that you are using as well?

**Reply by danielvan1** [reply]
To reproduce the error you can have a look at this GitHub repo:
https://github.com/danielvan1/GarnetExample

It creates 20k random arrays and writes them to the garnet server. I then try to do a vector sum for each of the 20k arrays in the `VectorSum.cs`.

To reproduce the error you can go to the root folder and:
1. Run `dotnet run --project Server/Server.csproj`
2. Run `dotnet run --project Client/Client.csproj`


The expected behavior is the error [described here](https://github.com/microsoft/garnet/discussions/665#discussioncomment-10735183).

**Reply by yrajas** [reply]
Thanks. Have the repro. 
Looking into potential solutions.
Thanks for your patience.

**Reply by yrajas** [reply, marked as answer]
Please check out the new APIs as part of this PR https://github.com/microsoft/garnet/pull/700 that help with buffer management like in your case.
